[DI - step 1] 코맥 미션 제출합니다.

parker1609 opened the pull request:
안녕하세요! 브리!

미션을 수행하면서 페어와 몇 가지 의문사항이 있어서 질문드립니다.

- 재귀 함수 리펙토링
`initializeInjectedBean2` 메서드가 리펙토링 전의 재귀함수인데, 현재 저희는 `initializeInjectedBean()` 메서드와 같이 메서드로 분리해서 리펙토링했습니다. 
그런데 이와 같이 하면 재귀 함수를 따라가기가 힘들다고 느껴졌습니다.
혹시 `initializeInjectedBean()`와 같이 분리하는 것이 좋은지, `initializeInjectedBean2()` 처럼 그대로 두는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

- assert 구문
`inject()` 메서드에서 `injectedConstructor` 변수에 대한 NULL체크를 `assert`를 통해 하고 있습니다. 그런데 이전에 호출되는 함수인 `initializeNotInjectedBeans()` 메서드에서 
이미 NULL체크를 하고 있습니다. 그러면 NULL체크를 두 번 해야 하는 상황인데, 이렇게 중복해서 체크하더라도 명시적으로 `assert`를 통해 체크하는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

부족하지만 잘 부탁드립니다! 감사합니다!

pkch93 committed:
docs: 요구사항 정리

pkch93 committed:
feat: BeanFactory initialize 메서드 구현

pkch93 committed:
refactor: BeanFactory.initialize() 메서드 리팩토링

pkch93 committed:
docs: 의문사항 작성

pkch93 committed:
refactor: typo 수정

pkch93 committed:
docs: todo 추가

pkch93 committed:
feat: BeanScanner 구현

pkch93 committed:
refactor: AnnotationHandlerMapping에 BeanFactory 주입

byjo commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+    @SuppressWarnings("unchecked")
     public void initialize() {
+        initializeNotInjectedBeans();
+        initializeInjectedBeans();
```
> inject() 메서드에서 injectedConstructor 변수에 대한 NULL체크를 assert를 통해 하고 있습니다. 그런데 이전에 호출되는 함수인 initializeNotInjectedBeans() 메서드에서
이미 NULL체크를 하고 있습니다. 그러면 NULL체크를 두 번 해야 하는 상황인데, 이렇게 중복해서 체크하더라도 명시적으로 assert를 통해 체크하는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

다른 객체에 의존관계가 없는 객체를 먼저 인스턴스화를 하셨군요.
그런데 제가 느끼기에는 `initializeNotInjectedBeans()`, `initializeInjectedBeans()` 두 메소드 내에서 행하는 처리는 동일한데 별개의 메소드로 분리되면서 흐름 파악이 더 어려워진 것 같아요.

두 메소드 모두 `preInstantiateBeans`에 대한 for문을 돌면서 아래의 공통된 로직이 있는데요.

1. 구현 클래스 조회 (`findConcreteClass()`)
2. @Inject 어노테이션이 있는 생성자 get (`BeanFactoryUtils.getInjectedConstructor()`) 및 null check
3. 인스턴스 생성(`getInstance()`) 및 map에 추가

재귀 함수 호출을 통해 필요한 객체를 먼저 인스턴스화 할 수 있는데, 메소드를 분리하신 이유가 있으실까요?

parker1609 replied:
피드백 감사합니다!

처음에 저희 생각은 생성자가 없는 경우(`injectedConstructor == null`)에는 기본 생성자를 통해 빈을 생성해주어야 하는데, 재귀 내부에서 이를 수행하면 예외가 발생할 수 있다고 생각한 것 같습니다. 리뷰어님의 피드백을 듣고 다시 한번 생각해보면서 재귀함수 내에서 처리할 수 있다는 것을 알 수 있었습니다. 따라서 재귀 함수 하나로 이를 처리하도록 변경하였습니다.

byjo commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+        }
+    }
+
+    private void initializeInjectedBean(Class clazz) {
```
> initializeInjectedBean2 메서드가 리펙토링 전의 재귀함수인데, 현재 저희는 initializeInjectedBean() 메서드와 같이 메서드로 분리해서 리펙토링했습니다.
그런데 이와 같이 하면 재귀 함수를 따라가기가 힘들다고 느껴졌습니다.
혹시 initializeInjectedBean()와 같이 분리하는 것이 좋은지, initializeInjectedBean2() 처럼 그대로 두는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

리팩토링 전의 `initializeInjectedBean2()` 메소드와 비교해보면 리팩토링을 통해 클린 코드의 원칙에 더 가까워졌다고 봐요. 
하나의 함수는 한 가지 일을 하고 있고, `initializeInjectedBean()` 메소드의 바디를 봤을 때 (구현의 구체적인 방법은 몰라도) 어떤 흐름으로 처리가 되는지 이해하기 더 쉬워졌네요.

다만, 말씀주신 것처럼 재귀 함수를 따라가기 어려운 것은 `initializeInjectedBean()`을 호출하는 곳에서 **빈으로 만들어진 클래스의 인스턴스는 `beans`라는 Map에 저장되어 있다** 라는 세부적인 내용을 알아야하기 때문이 아닐까 싶어요.

아래처럼 생성자의 인자로 들어가는 클래스의 인스턴스를 가져오기 위해서는 `beans` Map에 존재 여부를 확인하고 없으면 인스턴스를 생성한 후 Map에서 get을 해야 하네요.
https://github.com/woowacourse/jwp-di/blob/bd02df5d2603e7f1e7f1cc1ba799703278b4f826/nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java#L116-L119

`initializeInjectedBean2()` 메소드에서는 자신의 메소드를 재귀 호출하니 그 세부 로직이 보이지만, 리팩토링 후에는 메소드가 분리되었기 때문에 흐름 파악이 어렵지 않을까요?

`initializeInjectedBean()`이 이러한 세부 로직을 잘 포장하고, `initializeInjectedBean()` 메소드를 호출하는 쪽에서는 클래스에 맞는 인스턴스를 리턴 값으로 받을 수 있게 변경해보면 어떨까요?

parker1609 replied:
피드백 감사합니다!

리뷰어님 말씀을 듣고 `initializeInjectedBean()`, `prepareParameterBeans()` 두 군데에서 `if (!beans.containsKey(parameter))`을 하고 있는 것을 제거하면서 `initializeInjectedBean()`를 호출하는 쪽에 인스턴스를 리턴하도록 만들었습니다.

그런데 여기서 질문이 있습니다. 이렇게 리팩토링하면 인스턴스를 중복해서 만들게 될 것 같습니다.  그래서 가독성에서는 좋아졌지만 성능에 혹시 문제가 생기지 않을까 싶은데, 리뷰어님께서는 어떻게 생각하시는지 궁금합니다!

byjo commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+        }
+
+        Class concreteClass = findConcreteClass(clazz);
+        Object injectedBean = inject(concreteClass);
```
`inject()` 라는 메소드명이 해당 메소드를 잘 나타내고 있는지 약간 의문이 드네요. 
의미를 더 잘 드러낼 수 있는 메소드명을 고민해보면 어떨까요?
리팩토링을 하면서 분리한 메소드들도 행동에 맞는 이름을 가진다면 코드를 이해하는데 더 도움이 될 것 같아요.

parker1609 replied:
피드백 감사합니다!

`inject()` 가 수행하고 있는 로직은 매개변수로 전달받은 클래스 타입을 가지고 인스턴스를 생성하여 반환해주는 작업을 하고 있습니다. 이러한 로직을 나타내기 위해 `createInjectedInstance()` 라는 메서드명으로 변경하였습니다.

byjo commented on slipp/src/main/java/slipp/SlippWebApplicationInitializer.java:
```diff
@@ -17,8 +19,11 @@
 
     @Override
     public void onStartup(ServletContext servletContext) throws ServletException {
+        BeanFactory beanFactory = new BeanFactory(BeanScanner.scan("slipp"));
```
이번에 구현한 DI 프레임워크를 slipp 어플리케이션에서 사용하도록 변경하고, 잘 돌아가는지 확인해보면 어떨까요?

parker1609 replied:
피드백 감사합니다!

DM으로 자세한 설명 다시 한번 감사드립니다!ㅎㅎ `UserDao` 클래스에 `@Repository` 애노테이션을 선언한 뒤 빈으로 생성하여 주입하는 것으로 변경하였습니다.

byjo reviewed (changes requested):
코맥 안녕하세요~ 이번 미션 리뷰를 맡은 브리입니다.
DI 미션 하시면서 이것저것 많이 고민해보신 것 같아요. 👍
궁금하신 부분에 대한 제 생각을 공유드리니, 참고하셔서 리팩토링 해보면 좋을 것 같습니다.
혹시나 진행하다가 의문이 생기거나 이야기 나눠보고 싶으신 점 있다면 DM 주세요. 😊

pkch93 committed:
refactor: BeanFactory inject, getInstance 메서드명 변경

pkch93 committed:
refactor: BeanFactory initialize 메서드 중복제거

- initializeNotInjectedBeans 메서드가 initializeInjectedBeans와 같은 기능을
수행하고 있으므로 중복 부분 제거

pkch93 committed:
refactor: test 코드 package 구조 변경

pkch93 committed:
refactor: BeanScannerTest의 scanAll 메서드 DisplayName 추가

pkch93 committed:
docs: 추가 테스트 목록 작성

pkch93 committed:
test: BeanScanner 테스트케이스 추가

pkch93 committed:
docs: 의문사항 추가

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: di 프레임워크 적용하여 Controller, Dao 리펙토링

pkch93 committed:
refactor: AnnotationHandlerMapping

- addHandlerExecutions 메서드 불필요한 인자 제거
- BeanFactory.getControllers 메서드가 Set 반환하도록 변경

pkch93 committed:
refactor: BeanFactory.initialize 메서드에서 불필요한 bypass 제거

pkch93 committed:
test: BeanFactory

- 생성자 매개변수로 같은 종류의 어노테이션을 가진 빈을 생성하는 테스트 구현

pkch93 committed:
refactor: BeanFactory 클래스 중복제거

- beans.containKey 중복 제거

byjo commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,97 @@
+    }
+
+    private Object createBean(Class clazz) {
+        if (beans.containsKey(clazz)) {
+            return beans.get(clazz);
+        }
```
> 그런데 여기서 질문이 있습니다. 이렇게 리팩토링하면 인스턴스를 중복해서 만들게 될 것 같습니다. 그래서 가독성에서는 좋아졌지만 성능에 혹시 문제가 생기지 않을까 싶은데, 리뷰어님께서는 어떻게 생각하시는지 궁금합니다!

빈 생성(`createBean()`) 시 이미 존재하는 인스턴스인지 확인을 하는 로직이 있는데, 
인스턴스를 중복으로 만들 것으로 예상되는 부분이나 상황이 어디일까요?

byjo reviewed (approved):
피드백 반영 및 리팩토링 잘 하셨어요. 
이와 함께 테스트도 많이 보강하셨네요. 👏
질문 주신 사항에 대해서는 추가 설명 주시면 함께 고민해보겠습니다.
2단계 미션도 기대해볼게요!
